include/semaphore.h

```diff
@@ -113,7 +113,7 @@ struct sem_s
 # if CONFIG_SEM_PREALLOCHOLDERS > 0
   FAR struct semholder_s *hhead; /* List of holders of semaphore counts */
 # else
-  struct semholder_s holder[2];  /* Slot for old and new holder */
+  struct semholder_s holder;     /* Slot for old and new holder */
```

Comment:
hi @davids5 , I think there is no need to maintain two holders in the nxsem_restore_baseprio function, because when there is only one holder, nxsem_restore_baseprio only needs to restore the priority of the current task. Here, it is only necessary to select the highest priority based on the remaining holders of the current task, and there is no need to use the holder that has already been released.

Reply:
@zyfeier There has been a lot of resent changes to the priority restoration done by @pkarashchenko and the concern I have may have been resolved. But here is the [commit I was referring too](https://github.com/apache/nuttx/commit/360539afacc83132acdb83da8f20c468dbe4c63d)

It was an order of operation that required the change.

Priority inheritance: When CONFIG_SEM_PREALLOCHOLDERS==0, there is only a single, hard-allocated holder structure. This is problem because in sem_wait() the holder is released, but needs to remain in the holder container until sem_restorebaseprio() is called. The call to sem_restorebaseprio() must be one of the last things the sem_wait() does because it can cause the task to be suspended. If in sem_wait(), a new task gets the semaphore count then it will fail to allocate the holder and will not participate in priority inheritance. This fix is to add two hard-allocated holders in the sem_t structure: One of the old holder and one for the new holder.

Reply:
@davids5 the main idea here is in case of CONFIG_SEM_PREALLOCHOLDERS==0 not to wait until `sem_restorebaseprio()` call, but free holder immediately in `nxsem_release_holder()` so it can be directly reused. With assumption of `CONFIG_SEM_PREALLOCHOLDERS==0` the only task that requires priority restoration is the currently running task (the "old" holder).
I need to re-examine carefully if that's gonna work, but seems to be reasonable.